### DISCUSSION TITLE
Remove typos from asciicast recording

### DISCUSSION THREAD

**Discussion by matthieu-foucault** [asciinema/discussions#39] [Q&A]
Take this cast as an example: https://asciinema.org/a/165477
At 00:08, they type `git st`, then press backspace twice, to write `git log` instead.

A neat feature would be to allow to cleanup an existing .cast file by removing all of those typos, thus producing a smoother recording.

In this case, here are the related contents of the .cast file:

```
[7.52505,"o","g"]
[7.613211000000001,"o","i"]
[7.71004,"o","t"]
[7.7655460000000005,"o"," "]
[7.8933930000000005,"o","s"]
[7.995607000000001,"o","t"]
[8.684324,"o","\b\u001B[K"]
[8.851784,"o","\b\u001B[K"]
[9.184692,"o","l"]
[9.353251,"o","o"]
[9.572043,"o","g"]
```

The result of the cleanup would be to:
1. remove the `s`, `t`, and the two backspaces
2. offset all the subsequent times so that there is not a 1.5s gap in the recording

**Comment by ku1ik** [comment]
@matthieu-foucault this is something that I had on my mind for a while, and it's part of bigger feature family: asciicast post-processing.

I quickly hacked a script which you can use today for this: https://gist.github.com/sickill/934ac2148c1f5aa80fba4d983e646f67

It expects 2 arguments: original cast filename, and edited cast filename (that one will be created).

It converts timestamps to relative, saves it in tmp file, opens this tmp file in vim so you can edit or delete lines, and when you save and exit vim it converts timestamp back to absolute and saves it as proper v2 asciicast file :)

The input file can be either v2 or v1 asciicast, and the output is always v2. It doesn't copy over any header information though, so after running you need to copy width and height to the new file. And it doesn't preserve stdin events (if you recorded with `--stdin`). But I think it's already useful ;)

**Comment by ku1ik** [comment]
With latest changes in `develop` branch the following python script allows editing asciicast while preserving all information (header, stdin):

```python
import sys
import os
import asciinema.asciicast as asciicast
import asciinema.asciicast.v2 as v2
import asciinema.asciicast.events as ev


tmp_path = "/tmp/foo.cast"


with asciicast.open_from_url(sys.argv[1]) as cast:
    events = ev.to_relative_time(cast.events())

    with v2.writer(tmp_path, header=cast.v2_header) as w:
        for event in events:
            w.write_event(event)


os.system("vim " + tmp_path)


with asciicast.open_from_url(tmp_path) as cast:
    events = ev.to_absolute_time(cast.events())

    with v2.writer(sys.argv[2], header=cast.v2_header) as w:
        for event in events:
            w.write_event(event)


os.unlink(tmp_path)
```

You can run it from the directory where you check out `develop` branch.

**Comment by tyll** [comment]
I created similar tool. Do you know why the v1 format does not use relative timestamps anymore? It seems like a step backwards and I did not find a good reason why it would be better to use absolute timestamps.

**Comment by kaysond** [comment]
@sickill bump for a much needed feature! 
The last script you posted doesn't seem to be working for me. I get the following error:
```
Traceback (most recent call last):
  File "./edit-asciicast.py", line 14, in <module>
    with v2.writer(tmp_path, header=cast.v2_header) as w:
TypeError: __init__() got an unexpected keyword argument 'header'
```
